### DISCUSSION TITLE
Support more non-simultaneous axes via saving state in a plugin

### DISCUSSION THREAD

**Discussion by devtanc** [grblHAL/plugins#21] [Q&A] [answered]
Hey there!

I'm building a machine that has quite a number of axis and using a Teensy4.1 with the [grblhal breakout board](https://www.grbl.org/). I can drive 5 axis simultaneously with my current setup, and I've built a web UI for interfacing with the board over ethernet, which is working out great. I've also made a basic plugin for the board to allow for a custom M command set.

I have many more than just 5 axis, but never a need to **_simultaneously_** drive more than 5. I'm wondering if I could get a little bit of advice and direction in the idea of creating a plugin that would allow for an axis to be given a dual purpose (like an A1/A2 axis, where the two axis have identical function, but need to have their positions tracked separately). I'd like to be able to set the initial axis via a command like `M101Q___` where the Q value would be a bitfield enabling X1/X2 Y1/Y2 etc... When swapping between axis, it would save the "state" of the current 1 axis, then load the state of the 2 axis or vice versa.

I've already been doing this with a custom mux setup after the board that simply routes the signals to the correct axis, so I have that part working, but it leads to situations where my A axis thinks it's at 245, when the the one is at 100, and the other at 145. I'm just looking for a way to keep the positioning separate in the firmware via a plugin so that the data that I receive via the `?` command, and thus the positioning data the board has, shows the correct data given the currently selected axis.

My current idea is to access that position data and save the separate states in a plugin, which I'm conceptually familiar with, but just want some help as to how I can interface with the firmware to get and set that data at the right times without causing some real freak-outs for the board.

Any help is greatly appreciated!

**Comment by terjeio** [comment, marked as answer]
Do you call `sync_position()` to sync the parser and planner positions after modifying `sys.position`?
Since you have your own sender you can add the position of the swapped out axes to the real-time report by hooking into `grbl.on_realtime_report` event.

**Reply by devtanc** [reply]
I haven't yet written anything to modify or sync positions. Currently it just gradually gets more and more inaccurate.

From what you're saying, it sounds like I could just home the machine so that all the different axes are at a consistent starting position, then make a copy of `sys.position` that's local to my plugin. Then, when I swap an axis, I can just swap the relevant axis data currently in `sys.position` with my local copy to track both. And every time I do that, simply call `sync_position()` to make sure everything updates correctly?

And that would give me the most accurate positions so that things don't get out of sync with the actual machine position?

**Reply by terjeio** [reply]
Yes, try that.

**Reply by devtanc** [reply]
Just implemented this today and it's working like a dream! Thank you for the help!
